src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+        {
+            var generator = editor.Generator;
+            var matches = await FindMatchesAsync(document, expression, scope, allOccurrences, cancellationToken).ConfigureAwait(false);
+            var replacement = (TIdentifierNameSyntax)generator.IdentifierName(parameterName);
+
+            foreach (var match in matches)
+            {
+                // Special case the removal of the originating expression to either remove the local declaration
+                // or to add a rename annotation.
+                if (!match.Equals(expression))
+                {
+                    editor.ReplaceNode(match, replacement);
+                }
+                else
+                {
+                    if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName, out var localDeclaration))
+                    {
+                        editor.RemoveNode(localDeclaration);
+                    }
+                    else
+                    {
+                        // Creating a SyntaxToken from the new parameter name and adding an annotation to it
+                        // and passing that syntaxtoken in to generator.IdentifierName to create a new
+                        // IdentifierNameSyntax node.
+                        // Need to create the RenameAnnotation on the token itself otherwise it does not show up
+                        // properly.
+                        replacement = (TIdentifierNameSyntax)generator.IdentifierName(generator.Identifier(parameterName)
+                            .WithAdditionalAnnotations(RenameAnnotation.Create()));
+                        editor.ReplaceNode(match, replacement);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Goes through the parameters of the original method to get the location that the parameter
+        /// and argument should be introduced.
+        /// </summary>
+        private static int GetInsertionIndex(Compilation compilation, IMethodSymbol methodSymbol,
+            ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
+            Contract.ThrowIfNull(parameterList);
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in methodSymbol.Parameters)
+            {
+                // Want to skip optional parameters, params parameters, and CancellationToken since they should be at
+                // the end of the list.
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams &&
+                        !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
```

Comment:
can you move this method next to ShouldArgumentBeNamed.  they're effectively identical.